bpfman-operator/controllers/bpfman-agent/common.go

```diff
@@ -74,19 +74,234 @@ type ReconcilerCommon struct {
+	programList []client.Object) (ctrl.Result, error) {
+
+	r.Logger.V(1).Info("Start reconcileCommon(), programList:")
+	for _, program := range programList {
+		r.Logger.V(1).Info("    Program", "Name", program.GetName())
+	}
+
+	// Get existing ebpf state from bpfman.
+	programMap, err := bpfmanagentinternal.ListBpfmanPrograms(ctx, r.BpfmanClient, rec.getProgType())
+	if err != nil {
+		r.Logger.Error(err, "failed to list loaded bpfman programs")
+		return ctrl.Result{Requeue: true, RequeueAfter: retryDurationAgent}, nil
+	}
+
+	requeue := false // initialize requeue to false
+	for _, program := range programList {
+		r.Logger.V(1).Info("Reconciling program", "Name", program.GetName())
+
+		err := rec.setCurrentProgram(program)
+		if err != nil {
+			r.Logger.Error(err, "Failed to set current program")
+			return ctrl.Result{Requeue: true, RequeueAfter: retryDurationAgent}, nil
+		}
+
+		result, err := r.reconcileProgram(ctx, rec, program, programMap)
+		if err != nil {
+			r.Logger.Error(err, "Reconciling program failed", "Program Name", rec.getName, "ReconcileResult", result.String())
+		}
+
+		switch result {
+		case internal.Unchanged:
+			// continue with next program
+		case internal.Updated:
+			// return
+			return ctrl.Result{Requeue: false}, nil
+		case internal.Requeue:
+			// remember to do a requeue when we're done and continue with next program
+			requeue = true
+		}
+	}
+
+	if requeue {
+		// A requeue has been requested
+		return ctrl.Result{RequeueAfter: retryDurationAgent}, nil
+	} else {
+		// We've made it through all the programs in the list without anything being
+		// updated and a reque has not been requested.
+		return ctrl.Result{Requeue: false}, nil
+	}
+}
+
+// reconcileBpfmanPrograms ONLY reconciles the bpfman state for a single BpfProgram.
+// It does not interact with the k8s API in any way.
+func (r *ReconcilerCommon) reconcileBpfmanProgram(ctx context.Context,
+	rec bpfmanReconciler,
+	existingBpfPrograms map[string]*gobpfman.ListResponse_ListResult,
+	bpfProgram *bpfmaniov1alpha1.BpfProgram,
+	isNodeSelected bool,
+	isBeingDeleted bool,
+	mapOwnerStatus *MapOwnerParamStatus) (bpfmaniov1alpha1.BpfProgramConditionType, error) {
+
+	r.Logger.V(1).Info("enter reconcileBpfmanProgram()", "bpfProgram", bpfProgram.Name, "CurrentProgram", rec.getName())
+
+	uuid := bpfProgram.UID
+
+	noContainers := noContainersOnNode(bpfProgram)
+
+	existingProgram, doesProgramExist := existingBpfPrograms[string(uuid)]
+	if !doesProgramExist {
+		r.Logger.V(1).Info("Program doesn't exist on node", "CurrentProgram", rec.getName())
+
+		// If Program is being deleted just exit
+		if isBeingDeleted {
+			return bpfmaniov1alpha1.BpfProgCondUnloaded, nil
+		}
+
+		// Make sure if we're not selected just exit
+		if !isNodeSelected {
+			return bpfmaniov1alpha1.BpfProgCondNotSelected, nil
+		}
+
+		// If a container selector is present but there were no matching
+		// containers on this node, just exit.
+		if noContainers {
```

Comment:
hrm this comment is confusing me, as we're saying that it's checking "if container selector is present" but in reality it's only checking if there are "noContainers" maybe just update the `noContainersOnNode(` functions description?

Reply:
In getExpectedBpfPrograms(), if there is a container selector present, I see if I can find any matching containers on the given node.  If I can't find any matching containers, I set the UprobeNoContainersOnNode annotation. The noContainersOnNode() function returns true if that annotation exists and is set to true.  Does that make sense?  So, in effect, that's what it's telling us, but indirectly.  I can see if I can make the comments more clear.